### DISCUSSION TITLE
Is it idiomatic Teal to define all record methods up front?

### DISCUSSION THREAD

**Discussion by mtdowling** [teal-language/tl#773] [General]
I've been kicking the tires on Teal for a few weeks now, and it's great. One thing I'm not sure of is whether it's idiomatic to define all the methods of a record in the record definition, or if you only do that when you need forward declarations?

Do you do this:

```
local record Plant
    color: string
end

function Plant:grow()
    print("It's growing!")
end
```

Or this:

```
local record Plant
    color: string
    grow: function(self: Plant)
end

function Plant:grow()
    print("It's growing!")
end
```

**Comment by hishamhm** [comment]
Hi! Glad you're enjoying Teal! 

I don't think there is a consensus. We do need forward declarations when there are circular dependencies. This is a byproduct of the language implementation's simplicity, which is one of its tenets. Lua itself needs forward declarations with circular dependencies on local functions, so it's not that far off unexpected.

But I didn't want to enforce them in the language "for consistency" because I thought people would be annoyed by needless duplication (especially  for people coming from dynamic languages, every additional annotation you force on them feels like an extra burden).

So I would say it's fine to do it, and I don't think it feels "off" either (I think it's fine as at-a-glance documentation), but not necessary.

**Reply by parke** [reply]
hishamhm wrote:
> We do need forward declarations when there are circular dependencies. This is a byproduct of the language implementation's simplicity, which is one of its tenets.

@hishamhm - Would you be open to considering accepting a pull request that would remove the need to forward declare functions and methods?  I have enough familiarity with the general principles of compiler implementation that I believe I understand, in general, the work required to write such a pull request.  I am not that familiar with Teal's implementation, but I did look at Teal's code enough to see that it appears that the entire AST is fully parsed and available as an in-memory data structure that can be analyzed.

**Reply by hishamhm** [reply]
@parke Yes, it does build a full AST, but we currently do a single pass on it. As of now, Teal considers top-to-bottom declarations, with lexical scope: functions behave like any other variable. Instead of adding more ad-hoc rules for functions on top of Lua's that break the lexical scoping, I think a longer-term plan for avoiding the need for forward function declarations would be to rethink the global scope, and make it more like a proper module declaration and less like a Lua chunk. I don't have immediate plans to design and implement this, but it's not something that's out of the cards.

**Comment by Andre-LA** [comment]
I do like that for my scripts:

```lua
local record Plant
    color: string
end
-- optional, see [1]
local Plant_mt: metatable<Plant> = { __index = Plant }

function Plant.new(color: string): Plant
    return setmetatable({
        color = color,
    }, Plant_mt) -- [1]: could be just { __index = Plant }, thus no Plant_mt
end

function Plant:grow()
    -- ...
end

return Plant
```
